### DISCUSSION TITLE
Avalonia obfuscation

### DISCUSSION THREAD

**Discussion by MichaelSvejcar** [AvaloniaUI/Avalonia#12907] [Q&A]
Hello,
has any one from here tried some obfuscation tool in combination with Avalonia?
So far, I was trying ConfuserEx, but it was not working for some reason.
Thanks in advance!

**Comment by maxkatz6** [comment]
It works. 
NativeAOT and Obfuscar were confirmed to work with 11.0. 
Previous version of 0.10 required some additional configuration. 

> ConfuserEx, but it was not working for some reason.

It would be helpful to attach an error information and crossreport it to the ConfuserEx developers.

**Reply by MichaelSvejcar** [reply]
Thank you for the info.
I have tried updating to 11.0 and it's true that Obfuscar works with this version, however, there are so many breaking changes in my app, that it would probably take lot of time to update all successfully.
Because of this, do you know what additional configuration needs to be made, to make the Obfuscar working with 0.10 version? I was not able to find any information regarding this. 
Thanks in advance, enjoy the weekend!

**Reply by timunie** [reply]
your best bet is to enable compiled bindings everywhere. And also you need to add a roots.xml to your project. The themes need to be rooted iirc.

**Reply by Symbai** [reply]
It does not work with Eazfuscator (im using compiled bindings). Literally everything which is in XAML generates an error "Unable to resolve...". Eazfuscator does not officially support Avalonia but doesn't Avalonia convert XAML to IL so renaming shouldn't cause a problem?!

**Reply by maxkatz6** [reply]
It should work. There might be some problem with Eazfuscator then.
Or you use any API that might involve unreferenced code usage (like, StyleInclude in c# code, it has attributes and should produce trimming warnings as well, - StyleInclude in XAML is safe though). 

If any of these two is true, you probably can configure Eazfuscator to ignore XAML classes by the namespace of `CompiledAvaloniaXaml`.

**Reply by Symbai** [reply]
No nothing special. Even something simple like `Text={Binding Name}` failed as it couldn't find "Name" anymore. Or something like `Window.DataContext><vm:MyViewModel/>` it complaint it couldn't find "MyViewModel". This is what surprises me. It only worked if I exclude all classes and public properties from renaming but that is obviously not a solution. So I decided to try out AOT instead and I was surprised it worked out of the box. That is an acceptable alternative in my case.

**Reply by kekekeks** [reply]
> it complaint it couldn't find "MyViewModel"

It seems that your obfuscator somehow runs _before_ our XAML compiler. Run your obfuscator manually or at some later stage.

**Reply by Symbai** [reply]
That is not possible, Eazfuscator integrates into VS and runs at build/publish. It doesn't provide an option to obfuscate an already built assembly. But now that you mention it, it starts to make sense why it fails.

**Reply by MichaelSvejcar** [reply]
I successfully managed to run an Avalonia v0.10 project using Obfuscar with part by part obfuscation strategy. Initially, in the Obfuscar configuration file, I excluded all namespaces from the DLL I intended to obfuscate. Gradually, I reintroduced namespaces one at a time, continuously verifying the application's functionality after each addition.

In my experience, the application operated flawlessly when everything was obfuscated except for CompiledAvaloniaXaml, as well as namespaces associated with Views and ViewModels. If you aim to further obfuscate Views and ViewModels, I found that enabling compiled bindings, along with rebinding methods and commands in the code-behind instead in the view's XAML directly, proved to be a way to make it work.
